File: testing/unit-test-patterns/javascript/src/common/CloneEvaluationCentre.ts
import { IClone } from './interfaces/Clone.interface'
import { IUniqueCloneValidationRulesRepository } from './interfaces/UniqueCloneValidationRulesRepository.interface'
import { Gender } from './models/gender'
import { ValidationResult } from './models/ValidationResult'

export class CloneEvaluationCentre {
  constructor(
    private readonly uniqueCloneValidationRulesRepository: IUniqueCloneValidationRulesRepository
  ) {}

  async isCloneValid(clone: IClone): Promise<ValidationResult> {
    if (this.isGenericMale(clone) || this.isGenericFemale(clone)) {
      return ValidationResult.isValidResult()
    }

    const validFirstNames =
      this.uniqueCloneValidationRulesRepository.getValidFirstNames(clone.gender)
    const hasValidFirstName = validFirstNames.some(
      (x) => x.toLowerCase() === clone.firstName.toLowerCase()
    )
    if (!hasValidFirstName) {
      return ValidationResult.isInvalidResult('Clone has invalid first name')
    }

    const validLastNames =
      this.uniqueCloneValidationRulesRepository.getValidLastNames()
    const hasValidLastName = validLastNames.some(
      (x) => x.toLowerCase() === clone.lastName.toLowerCase()
    )
    if (!hasValidLastName) {
      return ValidationResult.isInvalidResult('Clone has invalid last name')
    }

    const maxNumberOfEyes =
      await this.uniqueCloneValidationRulesRepository.getMaximumNumberOfEyes()
    const hasValidNumberOfEyes = clone.numberOfEyes <= maxNumberOfEyes
    if (!hasValidNumberOfEyes) {
      return ValidationResult.isInvalidResult('Clone has too many eyes')
    }

    return ValidationResult.isValidResult()
  }

  private isGenericMale(clone: IClone): boolean {
    return (
      clone.firstName === 'John' &&
      clone.lastName === 'Smith' &&
      clone.gender === Gender.male &&
      clone.hairColor === 'Black' &&
      this.hasStandardNumberOfEyes(clone)
    )
  }

  private isGenericFemale(clone: IClone): boolean {
    return (
      clone.firstName === 'Jane' &&
      clone.lastName === 'Doe' &&
      clone.gender === Gender.female &&
      clone.hairColor === 'Blonde' &&
      this.hasStandardNumberOfEyes(clone)
    )
  }

  private hasStandardNumberOfEyes(clone: IClone): boolean {
    return clone.numberOfEyes === 2
  }
}

File: testing/unit-test-patterns/c#/UnitTestPatterns/Common/CloneEvaluationCentre.cs
using System;
using System.Linq;
using UnitTestPatterns.Common.Interfaces;
using UnitTestPatterns.Common.Models;

namespace UnitTestPatterns.Common
{
    public class CloneEvaluationCentre
    {
        private readonly IUniqueCloneValidationRulesRepository _uniqueCloneValidationRulesRepository;

        public CloneEvaluationCentre(IUniqueCloneValidationRulesRepository uniqueCloneValidationRulesRepository)
        {
            _uniqueCloneValidationRulesRepository = uniqueCloneValidationRulesRepository;
        }

        public ValidationResult IsCloneValid(IClone clone)
        {
            if (IsGenericMale(clone) || IsGenericFemale(clone))
            {
                return ValidationResult.IsValidResult();
            }

            var validFirstNames = _uniqueCloneValidationRulesRepository.GetValidFirstNames(clone.Gender);
            var hasValidFirstName = validFirstNames.Contains(clone.FirstName, StringComparer.InvariantCultureIgnoreCase);
            if (!hasValidFirstName)
            {
                return ValidationResult.IsInvalidResult("Clone has invalid first name");
            }
            
            var validLastNames = _uniqueCloneValidationRulesRepository.GetValidLastNames();
            var hasValidLastName = validLastNames.Contains(clone.LastName, StringComparer.InvariantCultureIgnoreCase);
            if (!hasValidLastName)
            {
                return ValidationResult.IsInvalidResult("Clone has invalid last name");
            }
            
            var maximumNumberOfEyes = _uniqueCloneValidationRulesRepository.GetMaximumNumberOfEyes();
            var hasCorrectNumberOfEyes = clone.NumberOfEyes <= maximumNumberOfEyes;
            if (!hasCorrectNumberOfEyes)
            {
                return ValidationResult.IsInvalidResult("Clone has too many eyes");
            }
            
            return ValidationResult.IsValidResult();
        }

        private static bool IsGenericMale(IClone clone)
        {
            return clone.FirstName == "John" &&
                   clone.LastName == "Smith" &&
                   clone.Gender == Gender.Male &&
                   clone.HairColor == "Black" &&
                   HasStandardNumberOfEyes(clone);
        }

        private static bool IsGenericFemale(IClone clone)
        {
            return clone.FirstName == "Jane" &&
                   clone.LastName == "Doe" &&
                   clone.Gender == Gender.Female &&
                   clone.HairColor == "Blonde" &&
                   HasStandardNumberOfEyes(clone);
        }

        private static bool HasStandardNumberOfEyes(IClone clone)
        {
            return clone.NumberOfEyes == 2;
        }
    }
}